Provide public API to determine if a symbol is accessible

gafter opened the pull request:
Fixes #24518
Fixes #23842

@dotnet/roslyn-compiler Please review this new proposed CodeAnalysis API
@dotnet/roslyn-ide I have removed the old Workspaces implementation (but not the tests)
@Pilchie @DustinCampbell @CyrusNajmabadi Feedback welcome.

CyrusNajmabadi commented:
Note: my only concern with this being at the compiler level is: does this accurately represent accessibility across both languages.  In the IDE it was fine to have a common 'good enough' impl that behaved reasonably well for C#/VB.  In other words, i never vetted it to make sure it was doing the right thing for VB, with the assumption that if it wasn't, it would likely be so corner case as to not matter for IDE features.

As a compiler API, i would like things to have a higher bar of accurately reflecting the language.  If this does that for VB and C#, then i'm happy :)

CyrusNajmabadi commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,291 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
```
i am very suprised github didn't see this as a move.  Did you change a lot of code in here?

gafter replied:
I didn't move it. I rewrote it from scratch to implement the language semantics.

CyrusNajmabadi replied:
> I didn't move it. I rewrote it from scratch to implement the language semantics.

Great!  Then that removes my concern about language semantics not matching.  Can you confirm these are the right semantics for VB and for C#?  Thanks!

CyrusNajmabadi commented:
>If VB differs, then it can't do both C# and VB at the same time.

Yeah... that would be the downside (and would block usage from the IDE, thus defeating hte purpose).

Basically, i'm ok with this if VB is ok with this.

CyrusNajmabadi commented:
> and CLR accessibility

That's a strong sign this would be ok.  Since the CLR is the common substrate between the languages, if this logic is at least providing those semantics, then i think this is fine no matter what.

gafter commented:
FYI, it looks like the VB implementation of accessibility mirrors the C# implementation, even when that conflicts with the language specification:

``` vb
            ' Protected is quite confusing. It's a two-fold test:
            '    1) The class access is from ("withinType") is, or is inside, 
            '        a class derived from "originalContainingType"
            '    2) If there is a qualifier, and the member is not shared, 
            '       the qualifier must be an instance of the derived
            '       class in which the access occurred (any construction thereof)
            '
            ' The VB language spec describes a third test:
            '    3) If there is a qualifier, and the member is not shared, the qualifier
            '       must be of the exact construction of the derived class from which access occurred.
            ' But Dev10 actually implements something rather different:
            '    3) If there is a qualifier, for any member, shared or not, the construction of "originalContainingType" that
            '       the qualifier type inherits from is the same as the construction of "originalContainingType" that the 
            '       class through which access occurred.
            ' This rule (either the spec'd or Dev10 version) is intentionally not implemented here. See bug 4107.
```

gafter commented:
@AlekseyTs I understand that you'd like more "verification" about compilations, but this API isn't intended to be bound to particular compilations (there are multiple symbol arguments and they can be from different compilations); compilations are not part of its contract. You say that the implementation as written "is likely to produce false negative results". Can you please identify specific scenarios in which it produces an incorrect result?

gafter commented:
test windows_release_vs-integration_prtest please

CyrusNajmabadi commented:
Note: the IDE creates symbols that do not belong to any compilation.  We do this as part of many features.  For example, we synthesize up methods during 'generate method' that are a real IMethodSymbol that are not backed by a C# or VB MethodSymbol.  These symbols then flow around many parts of our stack, including parts that want to potentially do accessibility checks.  

Having a language agnostic API allows us to pass this data around without having to ensure that all parts of the stack have to be agreeing ont he same language.

here's a good end-to-end example:

Say you have two projects, a C# and a VB project.  And you want to add 'add using with project reference'.  The search finds a potentially applicable symbol in teh VB project, but then wants to know if this symbol would be accessible at the location you're at in the C# project.  Note: we cannot 'map' the VB symbol into the C# project because there is literally no project reference between the two (that's what this feature is trying to decide if it should do).  So it wants to just say "would this symbol be accessible to the target language?" in an efficient manner.  i.e. we do not want to build up compilations, map symbols, and then ask queries in that fork when we can just ask a cheap question about two symbols.

gafter commented:
The intent of this API is to answer accessibility questions on `ISymbol`s, which contain enough information to answer accessibility questions. There is no need for us to delegate to language-specific symbol classes.

AlekseyTs commented:
> The intent of this API is to answer accessibility questions on ISymbols, which contain enough information to answer accessibility questions. There is no need for us to delegate to language-specific symbol classes.

I think there is no need to write new code, a third implementation of accessibility check, which actually doesn't provide benefits it might appear to provide. It looks like the implementation relies on symbol comparison, symbols from different languages are never equal. So, what is the point in having shared implementation that simply isn't going to have consistent behavior across all the scenarios it is expected to handle? If we don't want to put the method on a Compilation, we can simply add it as a real member on ISymbol interface, we add new APIs to symbol interfaces all the time. Why this one is special? Then, as with other similar APIs, we check that input belongs to the language and call implementation we already have. This, however, won't allow us to check whether the symbols belong to the same "universe". Note, that symbols that belong to the same language and represent the same entities might come from different "universes" and, therefore, will not be considered equal. A type symbol for System.Console obtained from one compilation might not be equal to symbol for System.Console obtained from another compilation. The shape of the API is likely to lead people to believe that it is not important where symbols are coming from, but it actually is extremely important. Some very simple scenarios, with public symbols, are likely to produce expected results, but the failures will actually come for more interesting scenarios. That said, I believe it is better to structure the API in the way that sets right expectations and leads consumers down the path of success rather than failure. Therefore, it feels bad to make it appear as though the API works reliably across languages, when it doesn't. That should be clearly documented and consumers should be prevented to call this API while mixing symbols from different languages. I would also prefer us to go farther and actually do more verification, making sure that symbols actually belong to the same compilation, only then we can guarantee correctness of the result across all scenarios.

AlekseyTs commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,291 @@
+                return true;
+            }
+
+            return toAssembly.GivesAccessTo(fromAssembly);
```
>GivesAccessTo [](start = 30, length = 13)

It looks like this is never true when toAssembly and fromAssembly belong to different languages.

gafter replied:
Sounds like a bug. Either an assembly in one language is permitted to give access to an assembly in another language, in which case the API is returning an incorrect result, or an assembly in one language is *not* permitted to give access to an assembly in another language, in which case this code is correct. Either way, if there is a problem it is in the implementation of `toAssembly.GivesAccessTo`, not here.

AlekseyTs replied:
> in which case this code is correct

Perhaps we should get clarity and agree on what does it mean for this component to behave correctly, what are the expected results in various scenarios. Your response makes me to believe that even you are not sure what to expect.

CyrusNajmabadi replied:
I agree it would be good to get clarity on all the relational cases.  This will help ensure everyone is on the same page and we all think it makes sense and can actually be implemented.  It will also be helpful toward driving a suitable test suite that then validates these expectations.  

From my reading of the code, the areas that need clarity are:

1. Internal, and how accessibilty across different-language-boundary should be defined.
2. protected, and how inheritance across different-language-boundary be defined.

Technically, it looks like 'private' also does type checks.  However, private seems well defined across languages/assemblies  It's always false as something in another assembly (which you would always get if you had a different language) would never be able to see your privates.

gafter replied:
> Your response makes me to believe that even you are not sure what to expect.

I am not sure whether there is a bug in the language-specific `Symbol` APIs, since you reported them not obeying their specifications. This API assumes that the Symbol APIs are correct.

I expect the `Symbol` APIs to return correct results, in which case this API would return correct results.

AlekseyTs replied:
>  am not sure whether there is a bug in the language-specific Symbol APIs, since you reported them not obeying their specifications.

Whether this is a bug or intentional designed behavior to keep language worlds separate is a separate discussion.

> I expect the Symbol APIs to return correct results, in which case this API would return correct results.

Where is a test for you assumptions and expectations? API do not have to behave the way you assume them to behave.

gafter replied:
I'll report a bug that the language-specific implementations do not obey their contract.

---
In reply to: [184278437](https://github.com/dotnet/roslyn/pull/26367#discussion_r184278437) [](ancestors = 184278437)

gafter replied:
> would you be willing to go investigate these and to come back with a proposal that we can discuss

Are you asking me to transcribe the CLR access rules? The only "variable" is how we define type and assembly identity. Assembly identity is supposed to be defined by` IAssemblySymbol.Identity` but @AlekseyTs points out that the compiler does it wrong. Type identity should be defined, recursively, by name and arity, and containing symbol. The rest is down to the CLR rules for the meaning of the access modes.

---
In reply to: [184489272](https://github.com/dotnet/roslyn/pull/26367#discussion_r184489272) [](ancestors = 184489272)

CyrusNajmabadi replied:
@gafter Thanks!

gafter replied:
We don't have tests for interfaces, because we don't have implementations in the interfaces.

---
In reply to: [184506449](https://github.com/dotnet/roslyn/pull/26367#discussion_r184506449) [](ancestors = 184506449)

gafter replied:
The code no longer compares assemblies using ==. We compare assembly identities instead, which is the CLR rule.

---
In reply to: [184566431](https://github.com/dotnet/roslyn/pull/26367#discussion_r184566431) [](ancestors = 184566431,184278437)

gafter replied:
The implementation now compares assemblies by comparing their identities.

---
In reply to: [184567028](https://github.com/dotnet/roslyn/pull/26367#discussion_r184567028) [](ancestors = 184567028,184489272)

CyrusNajmabadi commented:
> I do not believe the proposed implementation is going to give you all that.

Note, as part of the Neal did remove the near-equivalent version in teh IDE layer and replaced those existing calls with all the calls to his new API.  So it does appear to give us this sufficient level of accessibility testing.

> Therefore, it feels bad to make it appear as though the API works reliably across languages, when it doesn't. 

Note: the primary use case of this API was intended to support answering these questions for cross project scenarios.  That's how the IDE has been using the existing extension for many years now at this point.

> Also, keep in mind that this is a public API on symbols, the fault tolerance is much lower for it and correctness and usability requirements are much higher. 

I agree with this.

> making "promises" that are not fulfilled.

This part i'm less sure about, but i will leave neal to answer.  Specifically, the satisfactory part for me was that "It accurately reflects ... CLR accessibility."  If these are the rules for CLR accessibility, it feels like they can hopefully be exposed without actually needing or depending on specific symbol implementations.  If specific symbol implementations are necessary in your opinion, it would be good to understand where.

From your response on the PR one good example of a strange place is ```return toAssembly.GivesAccessTo(fromAssembly);```  It would be interesting to actually enumerate those cases and see if they can actually be written in a way that is truly symbol agnostic.  If so, then we can accomplish both sets of overall goals as far as i can tell.

CyrusNajmabadi commented:
> It looks like the implementation relies on symbol comparison, symbols from different languages are never equal.

That looks like a critique that doesn't torpedo the ship, but merely needs to be something addressed as part of the PR.  One question is: is it fundamental to accessibility checking that one be able to do symbol comparison?  Or can it be done without.  If the former, this appraoch may not be viable at all.  If the latter, that just indicates work to be done here to improve things.

Note: a concern of mine would be protected scenarios.   That's definitely a case that seems like it could be problematic.

>  Some very simple scenarios, with public symbols, are likely to produce expected results, but the failures will actually come for more interesting scenarios. 

IT would be interesting to enumerate those scenarios. That would help brainstorm to see if they really are deal-breakers, or if there are simple and effective solutions that could be created to validate that they are working properly.

AlekseyTs commented:
It looks like this PR lacks adequate testing and this is a show stopper on itself

gafter commented:
@AlekseyTs Not sure what you mean about testing - there is test coverage of every branch in the implementation. However, I will add a layer of testing based on mock symbols to ensure we do not depend on language-specific implementations.

AlekseyTs commented:
>  there is test coverage of every branch in the implementation

This on its own is not an indication of presence of adequate testing.

Here some of the things that I think should be present and is a good start:
- All scenarios should be covered by targeted tests under Compilers.
- All branches should be covered for all language specific symbols. Mock symbols might be useful, but this is lower in priority
- Cross language scenarios should be covered in all possible combinations. There are three inputs, the test matrix should be covered.
- Cross compilation scenarios should be covered when symbols are not shared between them, the same test matrix should be covered.

CyrusNajmabadi commented:
> Cross language scenarios should be covered in all possible combinations. There are three inputs, the test matrix should be covered.

I think this is a very good idea.  At the very least it will uncover bugs/issues elsewhere that we can drive solutions for.  For example, it would have found that our IAssembly symbols are not behaving as currently doc'ed.  

Even if these bugs have existed before, it would still be good to use this time as an opportunity to identify them and squash them so that we can have a veyr high quality implementation that people can know they can depend on.  

Note: we can discuss the necessity of fixing any bugs we find in other areas of the codebase if/when we find anything.  It could be acceptable to just file bugs on those parts and accept that those bugs have leaked into higher level APIs with the expectation that they could be fixed at some later point in time.  But it would still be good to know about them and have a good test suite that will appropriately hit those codepaths.

CyrusNajmabadi commented:
> Whether this is a bug or intentional designed behavior to keep language worlds separate is a separate discussion.

Note: my recollection of the time was this was not intentional.  Queries and other operations were intended to be supported across our common symbolic types.  The only thing we deliberately carved out were things we literally knew were impossible (or staggeringly difficult) to provide.  For example: INamedTypeSymbol.Construct.  There was just no way to provide these, without humongous refactorings of the compiler.  I argued at the time this method should go on Compilation (where things like Compilation.CreateArrayTypeSymbol and its ilk) live, to more clearly convey that you're deferring to the language to do this.  

However, the original design was simply that combining types in this fashion was basically the only thing that wasn't supported.  Querying with things from the ISymbol system was intended to be supported.

gafter commented:
This is intended to implement a function on the ISymbol API, not on the language-specific symbols. The mock tests demonstrate what features of the ISymbol API are and are not used, and how the API results affect the results of this API.

---
In reply to: [384761830](https://github.com/dotnet/roslyn/pull/26367#issuecomment-384761830) [](ancestors = 384761830)

CyrusNajmabadi commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,353 @@
+        private static bool SameModule(IModuleSymbol m1, IModuleSymbol m2)
+        {
+            Debug.Assert(m1 != null);
+            return m2 != null && SameOriginalSymbol(m1.ContainingSymbol, m2.ContainingSymbol);
```
modules have names (i think...).  Do those need to match?  If not, can you put a comment that it is intentional that we are not checking them?

gafter replied:
I don't think it can make any difference, since modules are merged into the assembly and cannot have conflicting members.


---
In reply to: [184568222](https://github.com/dotnet/roslyn/pull/26367#discussion_r184568222) [](ancestors = 184568222)

CyrusNajmabadi commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,353 @@
+            }
+
+            // We relax this to permit a type to be represented by multiple symbols (e.g. separate compilations)
+            return t2 != null && t1.Name == t2.Name && t1.Arity == t2.Arity && SameOriginalSymbol(t1.ContainingSymbol, t2.ContainingSymbol);
```
is direct name checking valid for VB cases?  i.e is case insensitivity a concern here?  Or is the expectation that that is not something between symbols.  If so, can you comment this is intentional.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,353 @@
+
+            if (!(within is INamedTypeSymbol || within is IAssemblySymbol))
+            {
+                throw new ArgumentException("must be an INamedTypeSymbol or an IAssemblySymbol", nameof(within));
```
since htis is a public API, i think these need to be loc'ed resources.  Right?

gafter replied:
We have mixed precedence. I'll localize.

---
In reply to: [184568430](https://github.com/dotnet/roslyn/pull/26367#discussion_r184568430) [](ancestors = 184568430)

CyrusNajmabadi commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,353 @@
+                case SymbolKind.PointerType:
+                    return IsAccessibleWithin(((IPointerTypeSymbol)symbol).PointedAtType, within, throughTypeOpt);
+                case SymbolKind.ErrorType:
+                    return true;
```
might be worth a comment

CyrusNajmabadi commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,353 @@
+                    return true;
+                case SymbolKind.Method:
+                    var method = (IMethodSymbol)symbol;
+                    if (((IMethodSymbol)symbol).MethodKind == MethodKind.BuiltinOperator)
```
double cast.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,353 @@
+            {
+                foreach (var typeArg in type.TypeArguments)
+                {
+                    if (!IsAccessibleWithin(typeArg, within, throughTypeOpt))
```
is passing throughTypeOpt to these recursive calls appropriate?  i would think the throughType was needed for chcking the original type.  It's lesss clear to me that it's right for checking the type arguments.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,353 @@
+            }
+            else
+            {
+                return IsMemberAccessible(containingType, type.DeclaredAccessibility, within, null);
```
nit: consider named arguments for null literals passed as args.  it helps make it clear what that parameter is for.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,353 @@
+
+            switch (declaredAccessibility)
+            {
+                case Accessibility.NotApplicable:
```
should this even be possible?  should we throw/assert for this?

gafter replied:
Yes, given that the caller filters out symbol kinds that would return this value.

---
In reply to: [184568746](https://github.com/dotnet/roslyn/pull/26367#discussion_r184568746) [](ancestors = 184568746)

CyrusNajmabadi commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,353 @@
+                return true;
+            }
+
+            var originalThroughTypeOpt = throughTypeOpt == null ? null : throughTypeOpt.OriginalDefinition as ITypeSymbol;
```
nit.  just use ```?.```?

CyrusNajmabadi replied:
i'm not sure whyt the "as ITypeSymbol" is there.  .OriginalDefinition should already be an ITypeSymbol

gafter commented:
@dotnet/roslyn-compiler May I please have a couple of reviews of this. All previous comments have been addressed.

gafter commented:
@DustinCampbell There are no VB-specific tests for the VB-specific access APIs to modify. All of the relevant VB tests are end-to-end language tests. The mock-based tests that are included here are language independent, and also included are cross-language tests.

gafter commented:
@CyrusNajmabadi I renamed the tests as you suggested, and moved private methods inside as local functions.

@dotnet/roslyn-compiler May I please have a couple of reviews of this new CompilerServices API?

gafter commented:
FYI, a future PR #26469 will fix bugs in the implementations of `IAssemblySymbol.GivesAccessTo` so that they do not give an incorrect result on assembly symbols coming from a different "universe". That is separate from this API addition, and until that (#26459) is fixed there are some cases that will not work correctly (tests included herein).

AlekseyTs commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,348 @@
+
+            bool hasInternalAccessTo(IAssemblySymbol fromAssembly, IAssemblySymbol toAssembly)
+            {
+                if (fromAssembly.Identity.Equals(toAssembly.Identity))
```
>fromAssembly.Identity.Equals(toAssembly.Identity) [](start = 20, length = 49)

I don't think this condition is appropriate to check for containment, however it appears that this is what it is used for.

gafter replied:
This is being used to determine if the two assembly symbols identify the same assembly.

gafter replied:
It is not being used for containment *per se*. It is used in the sense that if a program has both symbols loaded, then they are part of the same assembly and can access internals.

---
In reply to: [185161623](https://github.com/dotnet/roslyn/pull/26367#discussion_r185161623) [](ancestors = 185161623)

AlekseyTs replied:
> It is not being used for containment per se. It is used in the sense that if a program has both symbols loaded, then they are part of the same assembly and can access internals.

I argue that this logic is not correct. Equality of identities  says nothing about whether the assembly is the same.

---
In reply to: [185162836](https://github.com/dotnet/roslyn/pull/26367#discussion_r185162836) [](ancestors = 185162836,185161623)

AlekseyTs commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,348 @@
+                }
+            }
+
+            bool isNonNestedTypeAccessible(IAssemblySymbol assembly, Accessibility declaredAccessibility, ISymbol within0)
```
>assembly [](start = 59, length = 8)

Consider renaming to "declaringAssembly" for clarity

AlekseyTs commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,348 @@
+                }
+            }
+
+            bool hasInternalAccessTo(IAssemblySymbol fromAssembly, IAssemblySymbol toAssembly)
```
>IAssemblySymbol fromAssembly, IAssemblySymbol toAssembly) [](start = 37, length = 57)

The names of the parameters are confusing and I would say even somewhat backwards. It looks like we are going to check if ```toAssembly``` is giving access to ```fromAssembly```, it would be more intuitive if the names were reversed, I think.

gafter replied:
We're not checking whether one assembly *gives access to another*. We're checking if one assembly *has access to another*. Those are indeed reversed tests; and this test is for the latter, not the former as you surmise.

gafter replied:
I will rename the parameters for clarity.

---
In reply to: [185161731](https://github.com/dotnet/roslyn/pull/26367#discussion_r185161731) [](ancestors = 185161731)

AlekseyTs commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,348 @@
+                bool sameAssembly(IAssemblySymbol a1, IAssemblySymbol a2)
+                {
+                    Debug.Assert(a1 != null);
+                    return a2 != null && a1.Identity.Equals(a2.Identity);
```
>a1.Identity.Equals(a2.Identity) [](start = 41, length = 31)

I don't think this condition is appropriate to check for containment, but it appears this is what it is used for.

gafter replied:
It is being used to determine if the two assembly symbols identify the same assembly.

AlekseyTs replied:
> This is intended to align with the CLR's rules for identifying assemblies.

Please quote the section of the spec, the relevant wording used and provide explanation how it is applicable to detecting whether two type references are referencing the same definition.

---
In reply to: [185164217](https://github.com/dotnet/roslyn/pull/26367#discussion_r185164217) [](ancestors = 185164217,185124383)

AlekseyTs replied:
> It is being used to determine if the two assembly symbols identify the same assembly.

This is wrong, identity cannot answer this question.

---
In reply to: [185161761](https://github.com/dotnet/roslyn/pull/26367#discussion_r185161761) [](ancestors = 185161761)

gafter commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,348 @@
+                bool sameNamespace(INamespaceSymbol n1, INamespaceSymbol n2)
+                {
+                    Debug.Assert(n1 != null);
+                    // Note that the same symbol is expected to have the identical name as itself, despite VB language rules.
```
Your comment is too vague to be actionable. Please identify specific scenarios that you do not think will work well for cross-language queries. Please identify specific scenarios that you don't think will work well for VB to VB comparison across different compilations. Please identify specific scenarios that you believe might probably produce false positives.

AlekseyTs replied:
> Your comment is too vague to be actionable. Please identify specific scenarios that you do not think will work well for cross-language queries. Please identify specific scenarios that you don't think will work well for VB to VB comparison across different compilations. Please identify specific scenarios that you believe might probably produce false positives.

Clearly you already identified the inaccuracy of this check, hence the comment. VB namespaces are merged case-insensitively, it is obvious that the name of the namespace doesn't always reflect the actual casing of the name, therefore this comparison could produce both false negatives and false positives as the result. Since you are the author of the change, I leave it to you to come up with relevant unit-tests to demonstrate that. If you have trouble with those, I can help you offline with constructing scenarios.  

---
In reply to: [185161899](https://github.com/dotnet/roslyn/pull/26367#discussion_r185161899) [](ancestors = 185161899)

gafter replied:
Added a doc comment reflecting that, a test showing that, and filed an issue for the VB compiler not obeying the contract of `MetadataName`.

---
In reply to: [185126619](https://github.com/dotnet/roslyn/pull/26367#discussion_r185126619) [](ancestors = 185126619)

gafter commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,348 @@
+                // We expect a given named type definition to satisfy reference identity.
+                // But we relax this to permit a type to be represented by multiple symbols (e.g. separate compilations).
+                // Note that the same symbol is expected to have the identical name as itself, despite VB language rules.
+                return t1 == t2 || t2 != null && t1.Name == t2.Name && t1.Arity == t2.Arity && sameOriginalSymbol(t1.ContainingSymbol, t2.ContainingSymbol);
```
>t1.Name == t2.Name && t1.Arity == t2.Arity [](start = 49, length = 42)

To match the CLR rules, it is possible this should be `t1.MetadataName == t2.MetadataName`.

AlekseyTs commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,348 @@
+        /// an optional qualifier of type 'throughTypeOpt' to be used to resolve
+        /// protected access.
+        /// </summary>
+        public static bool IsAccessibleWithin(
```
>IsAccessibleWithin [](start = 27, length = 18)

Given the many "interesting" and controversial qualities of this API, I do not believe this API belongs in CodeAnalysis assembly as an extension method with such an innocent name. Compilers should not be using it under no circumstances in their internal implementation and since all symbols implement ISymbol, it is very easy to fall into this trap. Ideally this API should be pulled out into an assembly, which is not referenced by assemblies in the compiler layer, including this one.

gafter replied:
Added an Obsolete instance method to Symbol to interfere with the use of this extension method in the compilers.

---
In reply to: [185284396](https://github.com/dotnet/roslyn/pull/26367#discussion_r185284396) [](ancestors = 185284396)

AlekseyTs replied:
> Added an Obsolete instance method to Symbol to interfere with the use of this extension method in the compilers.

This is not sufficient, sometimes compilers operate with interfaces

---
In reply to: [185310868](https://github.com/dotnet/roslyn/pull/26367#discussion_r185310868) [](ancestors = 185310868,185284396)

gafter replied:
I am not aware of any scenarios where the compilers use interfaces when performing access checking.

---
In reply to: [185311211](https://github.com/dotnet/roslyn/pull/26367#discussion_r185311211) [](ancestors = 185311211,185310868,185284396)

AlekseyTs replied:
> I am not aware of any scenarios where the compilers use interfaces when performing access checking.

My concern was not limited to current code, and it is no addressed by your recent changes. Please do not mark this thread as resolved bases on that.

---
In reply to: [185311434](https://github.com/dotnet/roslyn/pull/26367#discussion_r185311434) [](ancestors = 185311434,185311211,185310868,185284396)

AlekseyTs commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,348 @@
+    public static partial class ISymbolExtensions
+    {
+        /// <summary>
+        /// Checks if 'symbol' is accessible from within 'within', with
```
>Checks if 'symbol' is accessible from within 'within', with [](start = 12, length = 59)

The doc comment sounds so innocent and doesn't provide any clues about inaccuracies built into this component and how they can affect the result. I think that at least all possibilities to get false positive and false negative results should be clearly enumerated with specific examples, etc. So that potential consumers clearly understand what they are getting into.

gafter replied:
I'm not aware of any false positive or false negative scenarios. Please identify *specific* scenarios that you theorize could be incorrect.

AlekseyTs replied:
> I'm not aware of any false positive or false negative scenarios. 

You are not aware or are you claiming there are not any?

---
In reply to: [185302839](https://github.com/dotnet/roslyn/pull/26367#discussion_r185302839) [](ancestors = 185302839)

gafter replied:
The code now uses metadata name to compare names, and .Equals to compare assemblies.

Please identify *specific* scenarios that you theorize could be incorrect, or else stop making this claim.


---
In reply to: [185303824](https://github.com/dotnet/roslyn/pull/26367#discussion_r185303824) [](ancestors = 185303824,185302839)

gafter commented:
One thing I should mention is that interactive and scripting implementations relax (i.e. ignore some of) the CLR rules for access checking. This implements the rules corresponding to what the compilers do.

---
In reply to: [384100119](https://github.com/dotnet/roslyn/pull/26367#issuecomment-384100119) [](ancestors = 384100119)

AlekseyTs commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,347 @@
+
+                bool sameAssembly(IAssemblySymbol a1, IAssemblySymbol a2)
+                {
+                    return a1.Equals(a2);
```
>a1.Equals(a2) [](start = 27, length = 13)

I believe, this code will not give accurate answers across languages

gafter replied:
I'll file a bug on the compilers.

---
In reply to: [185312547](https://github.com/dotnet/roslyn/pull/26367#discussion_r185312547) [](ancestors = 185312547)

AlekseyTs replied:
> I'll file a bug on the compilers.

This is definitely not a bug. This is by design

---
In reply to: [185313035](https://github.com/dotnet/roslyn/pull/26367#discussion_r185313035) [](ancestors = 185313035,185312547)

gafter replied:
Filed as #26542.

---
In reply to: [185313534](https://github.com/dotnet/roslyn/pull/26367#discussion_r185313534) [](ancestors = 185313534,185313035,185312547)

gafter commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,347 @@
+                {
+                    Debug.Assert(n1 != null);
+                    // Note that the same symbol is expected to have the identical name as itself, despite VB language rules.
+                    return n2 != null && n1.MetadataName == n2.MetadataName && sameOriginalSymbol(n1.ContainingSymbol, n2.ContainingSymbol);
```
Yes, I've filed an issue against the VB compiler.

gafter commented:
Adding the following doc comments:
``` c#
        /// <remarks>
        /// The following areas of imprecision may exist in the results of this API:
        /// <para>For assembly identity, we depend on Equals <see cref="IEquatable{T}.Equals(T)"/>/>.
        /// Assembly symbols that represent the same assembly imported into different language compilers
        /// may not compare equal, and this may prevent them from appearing to be related through
        /// this API. See https://github.com/dotnet/roslyn/issues/26542 .</para>
        /// <para>We compare <see cref="INamedTypeSymbol"/> based on the identity of the containing
        /// assembly (see above) and their metadata name, which includes the metadata name of the enclosing
        /// namespaces. Due to the behavior of the VB compiler (https://github.com/dotnet/roslyn/issues/26546)
        /// it merges namespaces when importing an assembly. Consequently, the metadata name of the namespace
        /// may not be correct for some of the contained types, and types that are distinct
        /// may appear to have the same fully-qualified name. In that case this API may treat them as the same type.</para>
        /// </remarks>
```

gafter commented:
An alternative would be to compare (original) type symbols using `object.Equals()` rather than comparing metadata names. That would change issue (2) in the doc comments.

gafter commented:
I will be changing the test for assembly identity to use `IAssembly.Identity.Equals`, since that is what is specified and used by the compilers and the runtime to determine when assemblies are the same.

gafter commented on src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs:
```diff
@@ -0,0 +1,367 @@
+                return withinType != null && isNestedWithinOriginalCDeclaringType(withinType, declaringType.OriginalDefinition);
+            }
+
+            bool isNestedWithinOriginalCDeclaringType(INamedTypeSymbol type, INamedTypeSymbol originalDeclatingType)
```
I think the extra `C` in the name of this local function is a typo.

gafter commented:
API review yesterday moved `IsAccessibleWithing` to `Compilation`. That is a large change from this API, so I'll close this and reopen a new PR.
